External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -185,52 +178,81 @@
         end
     else
         % Image contains only NaNs - skip this image
-    end
-
-
+	end
 end
 
-
 %% ========================================================================================
 %% ========================================================================================
 %% METHOD 3
-function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM, nEdgeVoxels)
+function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM, nEdgeVoxels, VoxelSize)
 %xASL_im_ExtrapolateLinearlyOverNaNs
 % The input of this function is 3D (iterations of other dimensions are done
 % in the main function above)
 %
 % IM          - input image
 % nEdgeVoxels - number of voxels at the edges of the non-nan values that we check for weird interpolation values, 
 % to be set to NaN and ignored in the extrapolation
+% VoxelSize   - voxel size is needed for extrapolating correctly as the transformation is given in mm and not voxels
 
 %% 0. Admin
 if nargin<2 || isempty(nEdgeVoxels)
     nEdgeVoxels = 0;
     % We can try this later, but it seems to work less well than doing the same based on the distance transform
 end
 
+if nargin < 3 || isempty(VoxelSize)
+	% The voxel-size is not provided and needs to be estimated from the transformation maps
+	% We set it here to empty in admin and estimate it later
+	VoxelSize = [];
+end
+
+%% Create masks and estimate voxel size
+% Create NaN mask for all 3D volumes together
+maskNaN = sum(isnan(IM), 4) > 0;
+
+% The voxel-size needs to be estimated
+if isempty(VoxelSize)
+	% For each volume, we calculate the difference of neighboring voxels
+	% We calculate the mean over nonNaN values, which gives the average update including the correct direction
+	diffIM = IM(2:end, :, :, 1) - IM(1:end-1, :, :, 1);
+	diffIM = diffIM(~maskNaN(2:end, :, :));
+	VoxelSize(1) = mean(diffIM(~isnan(diffIM)));
```

Comment:
Could replace this by xASL_stat_MeanNan.

Are you sure that the each volume only contains one dimension? Because that would be quite inefficient... that would mean that you could describe the flowfields in a single volume...